openmmtools/multistate/multistatesampler.py

```diff
@@ -1606,14 +1636,57 @@ def _is_completed_static(iteration_limit, iteration, last_err_free_energy,
     @staticmethod
     def _throw_restoration_error(message):
         """Masking function to hide the RestorationError class without exposing it or making it a 'hidden' (_X) error"""
+
         class RestorationError(Exception):
             """Represent errors occurring during attempts to restore simulations."""
+
             def __init__(self, error_message):
                 super().__init__(error_message)
                 # Critical messages which have halted a simulation, badly
                 logger.critical(error_message)
+
         raise RestorationError(message)
 
+    @staticmethod
+    def _create_context_caches(energy_context_cache_input, propagation_context_cache_input):
+        """Handle energy and propagation context cache default behavior.
+
+        .. note:: As of 01-Jan-22 default behavior is to use independent local caches.
+
+        Parameters
+        ----------
+        energy_context_cache_input : openmmtools.cache.ContextCache or None
+            Context cache to be used in energy computations. If None,
+            it will create a new unlimited ContextCache object.
+        propagation_context_cache_input : openmmtools.cache.ContextCache or None
+            Context cache to be used in the propagation of the mcmc moves. If None,
+            it will create a new unlimited ContextCache object.
+
+        Returns
+        -------
+        energy_context_cache : openmmtools.cache.ContextCache
+            Context cache to be used in energy computations.
+        propagation_context_cache : openmmtools.cache.ContextCache
+            Context cache to be used in the propagation of the mcmc moves.
+        """
+        # Handling energy context cache
+        if energy_context_cache_input is None:
+            # Default behavior, unlimited ContextCache
+            energy_context_cache = cache.ContextCache(capacity=None, time_to_live=None)
+        elif isinstance(energy_context_cache_input, cache.ContextCache):
+            energy_context_cache = energy_context_cache_input
+        else:
+            raise ValueError("Energy context cache input is not a valid ContextCache or None type.")
+        # Handling propagation context cache
+        if propagation_context_cache_input is None:
+            # Default behavior, unlimited ContextCache
+            propagation_context_cache = cache.ContextCache(capacity=None, time_to_live=None)
+        elif isinstance(propagation_context_cache_input, cache.ContextCache):
+            propagation_context_cache = propagation_context_cache_input
+        else:
+            raise ValueError("MCMC move context cache input is not a valid ContextCache or None type.")
+        return energy_context_cache, propagation_context_cache
```

Comment:
This method is internal, so it can just store these as `self.energy_context_cache` and `self.propagation_context_cache`. It doesn't need to return them.

Reply:
Here I was just thinking about explicitly setting the attributes (even if private/internal) in the `__init__`, because, as far as I know, it is commonly frowned upon to set them in another method, even if the method is called in the `__init__` I think it makes it a little cryptic. What do you think?

Reply:
That's fine---I agree your approach is less confusing.